resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/builders/AutoBuildJobTest.java

```diff
@@ -139,11 +153,87 @@ public IProject[] build(int kind, Map<String, String> args, IProgressMonitor mon
 			}
 		});
 
-		// triggers autobuild
-		project.touch(getMonitor());
-		Thread.sleep(Policy.MAX_BUILD_DELAY);
+		triggerAutoBuildAndWait();
 		Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
 		assertEquals("Should see two scheduled() calls", 2, scheduled.get());
 		assertEquals("Should see two running() calls", 2, running.get());
 	}
-}
+
+	@Test
+	public void testWaitForAutoBuild_JobManagerIsSuspended_ExceptionIsThrown()
+			throws InterruptedException, CoreException, ExecutionException {
+		Job.getJobManager().suspend();
+
+		assertEquals("Scheduled calls", 0, scheduled.get());
+		assertEquals("Running calls", 0, running.get());
+
+		triggerAutoBuildAndWait();
+		Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
+
+		assertEquals("Scheduled calls", 1, scheduled.get());
+		assertEquals("Running calls", 0, running.get());
+
+		Throwable thrown = waitForAutoBuild(2_000);
+		assertTrue(JobManagerSuspendedException.class.isInstance(thrown));
+
+		Job.getJobManager().resume();
+	}
+
+	@Test
+	public void testWaitForAutoBuild_JobManagerIsRunning_NoExceptionIsThrown()
+			throws InterruptedException, CoreException, ExecutionException {
+		// make sure the auto-build can even happen
+		Job.getJobManager().resume();
+
+		assertEquals("Scheduled calls", 0, scheduled.get());
+		assertEquals("Running calls", 0, running.get());
+
+		triggerAutoBuildAndWait();
+		Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
+
+		assertEquals("Scheduled calls", 1, scheduled.get());
+		assertEquals("Running calls", 1, running.get());
+
+		Throwable thrown = waitForAutoBuild(2_000);
+		assertNull(thrown);
+	}
+
+	/**
+	 * Trigger an auto-build and wait for it to start.
+	 *
+	 * @throws CoreException
+	 * @throws InterruptedException
+	 */
+	private void triggerAutoBuildAndWait() throws CoreException, InterruptedException {
+		project.touch(getMonitor());
+		Thread.sleep(Policy.MAX_BUILD_DELAY);
+	}
+
+	/**
+	 * Wait for an auto-build operation to finish.
+	 *
+	 * @param timeoutMillis
+	 *            after this timeout, a <code>TimeoutException</code> will be
+	 *            thrown.
+	 * @return the exception thrown by {@linkplain BuildManager#waitForAutoBuild()},
+	 *         it can be <code>null</code>.
+	 * @throws InterruptedException
+	 *             if the thread waiting for the auto-build is interrupted.
+	 */
+	private Throwable waitForAutoBuild(long timeoutMillis) throws InterruptedException {
+		try {
+			ForkJoinPool.commonPool()//
+					.submit(() -> getBuildManager().waitForAutoBuild())//
+					.get(timeoutMillis, TimeUnit.MILLISECONDS);
+		} catch (ExecutionException e) {
+			// Since the wait is happening in a Future, the original exception is packed
+			// inside the ExecutionException
+			return e.getCause();
```

Comment:
I'd propose to have the method rethrow the actual exception to be checked for instead of returning the exception object. Then you do not have to return `null` as a marker and can use a proper `assertThrows` in the calling method.

```suggestion
			throw e.getCause();
```

Reply:
Good idea, thank you! I just changed it

Reply:
The method should be changed to `void` return type.